### ISSUE TITLE
Tailwind JIT doesn't work with Next.js

### ISSUE THREAD

**Issue by BorisZubchenko** [nrwl/nx#6332]
## Current Behavior

When Tailwind JIT mode is enabled, applying Tailwind rules within css files only works:
```css
h1 {
  @apply bg-[#fef]
}
```
However, it is impossible to use Tailwind rules within `tsx` files **at all**:
```tsx
export function Index() {
  return <>
    <h1 className='bg-[#fef]'>Rules hasn't been applied!</h1>
    <h1 className='font-bold'>Rules hasn't been applied!</h1>
  </>;
}

export default Index;
```

P.S. I generated a Next.js application without NX and it works. Seems like it's a bug in the executor.

## Expected Behavior

Tailwind rules have to be applied with the JIT mode enabled.

## Steps to Reproduce

1. `npx --ignore-existing create-nx-workspace happynrwl --preset=next`
2. `cd happynrwl`
3. `npm install -D tailwindcss@latest postcss@latest autoprefixer@latest`
4. `cd apps/frontend`
5. `npx tailwindcss init -p`
6. Edit `postcss.config.js`:
```js
module.exports = {
  plugins: {
    tailwindcss: { config: './apps/frontend/tailwind.config.js' },
    autoprefixer: {},
  },
};
```
7. Edit `tailwindcss.config.js`:
```js
const { createGlobPatternsForDependencies } = require('@nrwl/react/tailwind');
module.exports = {
  mode: 'jit'
  purge: createGlobPatternsForDependencies(__dirname),
  // ...
};
```
8. Add the following lines to `styles.css`:
```
@tailwind base;
@tailwind components;
@tailwind utilities;

h1 {
  @apply bg-[#0ff]
}
```
9. Add some tags to `index.tsx`:
```
export function Index() {
  return <>
    <h1>Red title works!</h1>
    <h2 className='bg-green-400'>Green title doesn't work!</h2>
  </>;
}
10. `npm run start`
```

### Environment
```
 Node : 14.17.3
  OS   : linux x64
  npm  : 6.14.13
  
  nx : Not Found
  @nrwl/angular : Not Found
  @nrwl/cli : 12.5.7
  @nrwl/cypress : 12.5.7
  @nrwl/devkit : 12.5.7
  @nrwl/eslint-plugin-nx : 12.5.7
  @nrwl/express : Not Found
  @nrwl/jest : 12.5.7
  @nrwl/linter : 12.5.7
  @nrwl/nest : Not Found
  @nrwl/next : 12.5.7
  @nrwl/node : Not Found
  @nrwl/react : 12.5.7
  @nrwl/schematics : Not Found
  @nrwl/tao : 12.5.7
  @nrwl/web : 12.5.7
  @nrwl/workspace : 12.5.7
  @nrwl/storybook : 12.5.7
  @nrwl/gatsby : Not Found
  typescript : 4.2.4

```

**Comment by juristr** [comment]
Thanks for mentioning this issue. Can you try to adjust the `purge` field? Maybe this is a misunderstanding, but the `createGlobPatternsForDependencies` only creates the glob patterns for dependencies of the app you're having the `tailwind.config.js` within.

You should have a purge pattern similar to what [I mentioned in my article](https://juristr.com/blog/2021/06/setup-tailwind-nextjs-and-nx/). Here's the code:

https://github.com/juristr/blog-series-nextjs-nx/blob/1e2a519cc7cb54c0306e385c2ef48fadd554626d/apps/site/tailwind.config.js#L5-L9

Basically add the glob patterns for the app and then add the `createGlobPatternsForDependencies` portion for potential dependencies (which I get you don't have since you don't have any libs in the demo project).

**Comment by BorisZubchenko** [comment]
I've just tested my demo with the purge config you provided. Tailwind rules apply, but only for the first time:

- I apply `bg-red-400`
- Serve the application
- The element is indeed `red`
- After this, I apply `bg-green-400`
- The element is **transparent**
- After this, I apply `bg-red-400`
- The element is `red` again

Seems like an AOT, not JIT 😅. Unless I rebuild the app every time I do the change.

**Comment by BorisZubchenko** [comment]
I saw the following point:

> `TAILWIND_MODE` to `watch` to get continuous watching as you have the dev server running and change the Tailwind classes. Set it to `build` for a one-off compilation

So I've made it working:
```
TAILWIND_MODE=watch nx run site:serve
```

However, I think it can be done at the executor level, so we set `TAILWIND_MODE=watch` while in non-prod env.

**Comment by juristr** [comment]
Yeah that watch Environment variable is needed right now. We probably wouldn't want to add this to the Next executor because this is rather very specific to tailwind itself.

My goto solution for this is usually to create a NPM script in package.json or add a run-command executor in the workspace.json to launch the app. Makes sense?
I can post an example later / tomorrow (on my mobile rn)

**Comment by BorisZubchenko** [comment]
An example of the executor would be great, thanks!

**Comment by BorisZubchenko** [comment]
Btw, I've just set up Storybook and had to set `TAILWIND_MODE=watch` as well. I'm thinking about hoisting this kind of logic to the executor's level. For now, I have to edit IDE's run configurations to add envs. However, my `.idea` folder is git ignored, so this approach is not "portable".

**Comment by nartc** [comment]
Just want to chime in a little bit about adding the Env Var to the executor, the AngularCLI builder assigns value to TAILWIND_MODE if it detects the project is using Tailwind (based on `tailwind.config.js`) 
https://github.com/angular/angular-cli/blob/fe25ab1ef2ce1474db22217a3bdc36db03c18978/packages/angular_devkit/build_angular/src/webpack/configs/styles.ts#L159

But I guess Next executor doesn't try to detect anything Tailwind specific since the consumers own the postcss config in Next project.

**Comment by juristr** [comment]
Thanks folks. Let me have a look at this and see how/whether we can improve the experience here. Won't be able this week though. Next week's blog post for the Next + Nx + Tailwind series will be about Storybook and Tailwind. Might be next or the week after. I have to first do some adjustments in the Nx src.

I'll try to look into this as part of that work. Will ping you for further clarifications 🙂

**Comment by BorisZubchenko** [comment]
> Thanks folks. Let me have a look at this and see how/whether we can improve the experience here. Won't be able this week though. Next week's blog post for the Next + Nx + Tailwind series will be about Storybook and Tailwind. Might be next or the week after. I have to first do some adjustments in the Nx src.
> 
> I'll try to look into this as part of that work. Will ping you for further clarifications

Thanks! In turn, feel free to ask for help with your blog post 😉. I've just set up Next+Nx+Tailwind+Storybook, had to spent some time on it though.

**Comment by bogdansoare** [comment]
Also having this issue, on initial page load the styles load, but after that if I change something the JIT doesn't work

**Comment by juristr** [comment]
@BorisZubchenko Thanks! In turn, feel free to ask for help with your blog post 😉. I've just set up Next+Nx+Tailwind+Storybook, had to spent some time on it though.

Definitely will. I have some changes to push into Nx before writing that blog post which will make the dev ergonomics better for the React Storybook setup in Nx. Wanted to push it for v12.6, but didn't make it, so it's gonna go into 12.7. Thus the Storybook blog post will be delayed a bit. I might push another one first meanwhile :)

@bogdansoare You can prefix it with the `TAILWIND_WATCH=true` env variable for now as mentioned [here](https://github.com/nrwl/nx/issues/6332#issuecomment-878509542) and also on the blog post.

I'll def look into trying to come up with a better DX for this

**Comment by nphivu414** [comment]
> purge pattern

Thanks for pointing this, after adding the purge pattern, JIT mode works great for me

**Comment by juristr** [comment]
@BorisZubchenko You mentioned having setup Nx + Tailwind with Storybook. Do you have a repo to share? I'd like to have quick look to see how you set it up and check difference / optimization points as I'm preparing to merge some React related Storybook improvements. Thanks!
(you can also DM me on Twitter or Nrwl Community Slack)

**Comment by BorisZubchenko** [comment]
@juristr, have a look at the following PR: https://github.com/BorisZubchenko/mono/pull/27

It would be great if you could mention some issues in it 🙂

**Comment by marckassay** [comment]
@juristr You mentioned, "...You can prefix it with the **TAILWIND_WATCH=true** env variable...". I assume you meant `TAILWIND_MODE=watch`?

**Comment by juristr** [comment]
@marckassay yep, ofc 😅

**Comment by NikitaIT** [comment]
I honestly don't know why it works. But it seems to have completely solved the problem.

````ts
// <sub-root>/.storybook/main.js
webpackFinal: async (config, { configType }) => {
    if (process.env.NODE_ENV === 'development') {
      process.env.TAILWIND_MODE = 'watch';
    }
}

// <sub-root>/.storybook/preview.js
import '../src/styles/main.css'; /* path to tailwind imports */

// <sub-root>/postcss.config.js
require('tailwindcss')({
      config: join(__dirname, 'tailwind.config.js'), 
      /* It doesn't find a config without an explicit declaration. The postcss storybook plugin also has this problem. */
      /* More precisely, I know why. It always searches from root, apparently from the location of node-modules or the nearest pkg.json which is not in nx packages by default. */
}),
```

**Comment by BorisZubchenko** [comment]
I've just checked it once again it it seems to work:

postcss.config.js
```javascript
const { join } = require('path')

module.exports = {
  plugins: {
    tailwindcss: {
      config: join(__dirname, 'tailwind.config.js'),
    },
    autoprefixer: {},
  },
}

```

tailwind.config.js:
```javascript
const { join } = require('path')
const { createGlobPatternsForDependencies } = require('@nrwl/next/tailwind')

module.exports = {
  presets: [require('../../tailwind-workspace-preset.js')],
  content: [
    join(__dirname, 'pages/**/*.{js,ts,jsx,tsx}'),
    ...createGlobPatternsForDependencies(__dirname),
  ],
  mode: 'jit',
  theme: {
    extend: {},
  },
  plugins: [],
}

```

I don't know what has changed, maybe the `purge` option became `content` and some Tailwind internals changed. But now I can confirm that JIT works!

```
  Node : 14.17.3
  OS   : linux x64
  npm  : 6.14.13
  
  nx : 13.3.6
  @nrwl/angular : undefined
  @nrwl/cli : 13.3.6
  @nrwl/cypress : 13.3.6
  @nrwl/devkit : 13.3.6
  @nrwl/eslint-plugin-nx : 13.3.6
  @nrwl/express : undefined
  @nrwl/jest : 13.3.6
  @nrwl/linter : 13.3.6
  @nrwl/nest : undefined
  @nrwl/next : 13.3.6
  @nrwl/node : undefined
  @nrwl/nx-cloud : 13.0.0
  @nrwl/react : 13.3.6
  @nrwl/react-native : undefined
  @nrwl/schematics : undefined
  @nrwl/tao : 13.3.6
  @nrwl/web : 13.3.6
  @nrwl/workspace : 13.3.6
  @nrwl/storybook : 13.3.6
  @nrwl/gatsby : undefined
  typescript : 4.4.4
  rxjs : 6.6.7
  tailwindcss: 3.0.5
  postcss: 8.4.5
  autoprefixer: 10.4.0
```
